library/std/src/sys/unix/locks/queue_rwlock.rs

```diff
@@ -0,0 +1,492 @@
+            // Since at least one element in the queue has a non-null tail (invariant 3.),
+            // this code will never be run for `current == tail`.
+            None => unsafe {
+                let next = c.next.get().unwrap_unchecked();
+                next.as_ref().prev.set(Some(current));
+                current = next;
+            },
+        }
+    };
+
+    unsafe {
+        head.as_ref().tail.set(Some(tail));
+        tail
+    }
+}
+
+pub struct RwLock {
+    state: AtomicState,
+}
+
+impl RwLock {
+    #[inline]
+    pub const fn new() -> RwLock {
+        RwLock { state: AtomicPtr::new(UNLOCKED) }
+    }
+
+    #[inline]
+    pub fn try_read(&self) -> bool {
+        self.state.fetch_update(Acquire, Relaxed, lock(true)).is_ok()
+    }
+
+    #[inline]
+    pub fn read(&self) {
+        if !self.try_read() {
+            self.lock_contended(true)
+        }
+    }
+
+    #[inline]
+    pub fn try_write(&self) -> bool {
+        // This is lowered to a single atomic instruction on most modern processors
+        // (e.g. "lock bts" on x86 and "ldseta" on modern AArch64), and therefore
+        // is more efficient than `fetch_update(lock(false))`, which can spuriously
+        // fail if a new node is appended to the queue.
+        self.state.fetch_or(LOCKED, Acquire).addr() & LOCKED != LOCKED
+    }
+
+    #[inline]
+    pub fn write(&self) {
+        if !self.try_write() {
+            self.lock_contended(false)
+        }
+    }
+
+    #[cold]
+    fn lock_contended(&self, read: bool) {
+        let update = lock(read);
+        let mut node = Node::new(read);
+        let mut state = self.state.load(Relaxed);
+        let mut count = 0;
+        loop {
+            if let Some(next) = update(state) {
+                // The lock is available, try locking it.
+                match self.state.compare_exchange_weak(state, next, Acquire, Relaxed) {
+                    Ok(_) => return,
+                    Err(new) => state = new,
+                }
+            } else if count < SPIN_COUNT {
+                // If the lock is not available, spin for a while.
+                spin_loop();
+                state = self.state.load(Relaxed);
+                count += 1;
+            } else {
+                // Fall back to parking. First, prepare the node.
+                node.prepare();
+                node.set_state(state);
+                node.prev = AtomicLink::new(None);
+                // If this is the first node in the queue, set the tail field to
+                // the node itself to ensure there is a current `tail` field in
+                // the queue (invariants 3. and 4.). This needs to use `set` to
+                // avoid invalidating the new pointer.
+                node.tail.set((state.addr() & QUEUED == 0).then_some(NonNull::from(&node)));
+
+                let next = ptr::from_ref(&node)
+                    .map_addr(|addr| addr | QUEUED | state.addr() & LOCKED)
```

Comment:
Operator precedence of the & is a bit confusing here. Its probably better to just | LOCKED instead of using .addr() on the old state.

Reply:
I'll use braces. The lock may be unlocked at this point if a writer was last in the queue when the lock was unlocked.